designcompose/src/main/java/com/android/designcompose/squoosh/SquooshRender.kt

```diff
@@ -248,13 +248,30 @@ private fun squooshTextRender(
     nodeName: String,
     variableState: VariableState,
 ) {
-    val paragraph =
+    var paragraph =
         Paragraph(
             paragraphIntrinsics = textInfo.paragraph,
             width = computedLayout.width * density.density,
             maxLines = textInfo.maxLines,
             ellipsis = style.node_style.text_overflow is TextOverflow.Ellipsis
         )
+    // Clip canvas will clip the shadow too. Not sure if there is other better ways to compute the
+    // max lines manually.
+    val bottom = computedLayout.height * density.density
+    val cutLine = paragraph.getLineForVerticalPosition(bottom)
+    val cutLineBottom = paragraph.getLineBottom(cutLine)
+    if (cutLineBottom > bottom) {
+        val maxLines = cutLine.coerceAtLeast(1)
+        if (maxLines != textInfo.maxLines) {
+            paragraph =
```

Comment:
I think this should take a fast path through `Paragraph` because the width isn't changing, so it might be OK to construct a paragraph here during render rather than during layout -- but please do see if it's possible to do this during layout instead. We should try to get this change in as it's a big improvement in text compatibility.

Reply:
I looked into how the layout is done but I am not sure if I get it right.
The node in the figma file has a fixed width `291` and a fixed height `61`. The size is passed to taffy to do the measurement which will call the https://github.com/google/automotive-design-compose/blob/170257c7a91c2de2335878b337b09e5424c5654d/crates/layout/src/layout_manager.rs#L183 which then calls the kotlin https://github.com/google/automotive-design-compose/blob/170257c7a91c2de2335878b337b09e5424c5654d/designcompose/src/main/java/com/android/designcompose/DesignText.kt#L482 to do the text measurement.

However, the size passed into the measure function for the height though is 0.0 with AvailableHight to be MinContent so I got 0s for the height and available height in the kotlin measure code https://github.com/google/automotive-design-compose/blob/170257c7a91c2de2335878b337b09e5424c5654d/designcompose/src/main/java/com/android/designcompose/DesignText.kt#L482 where I think is the best place to calculate how many lines the layout could fit.

I will check with @rylin8 tomorrow when we are in the office to see if he can help me figure out more about it.

Reply:
I changed to do the canvas clipping approach because of max lines can not cut off part of the line:
<img width="730" alt="Screenshot 2024-08-20 at 15 58 38" src="https://github.com/user-attachments/assets/852de212-02b4-44bd-9c6b-35f1063dd9a8">